test/Dialect/FIRRTL/canonicalization.mlir

```diff
@@ -1530,4 +1531,51 @@ firrtl.module @issue1139(out %z: !firrtl.uint<4>) {
 }
 
+// CHECK-LABEL: firrtl.module @PadMuxOperands
+firrtl.module @PadMuxOperands(
+  in %cond: !firrtl.uint<1>,
+  in %ui: !firrtl.uint,
+  in %ui11: !firrtl.uint<11>,
+  in %ui17: !firrtl.uint<17>,
+  out %z: !firrtl.uint
+) {
+  %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+  %c1_ui1 = firrtl.constant 1 : !firrtl.uint<1>
+
+  // Smaller operand should pad to result width.
+  // CHECK: %0 = firrtl.pad %ui11, 17 : (!firrtl.uint<11>) -> !firrtl.uint<17>
+  // CHECK: %1 = firrtl.mux(%cond, %0, %ui17) : (!firrtl.uint<1>, !firrtl.uint<17>, !firrtl.uint<17>) -> !firrtl.uint<17>
+  // CHECK: %2 = firrtl.pad %ui11, 17 : (!firrtl.uint<11>) -> !firrtl.uint<17>
+  // CHECK: %3 = firrtl.mux(%cond, %ui17, %2) : (!firrtl.uint<1>, !firrtl.uint<17>, !firrtl.uint<17>) -> !firrtl.uint<17>
+  %0 = firrtl.mux(%cond, %ui11, %ui17) : (!firrtl.uint<1>, !firrtl.uint<11>, !firrtl.uint<17>) -> !firrtl.uint<17>
+  %1 = firrtl.mux(%cond, %ui17, %ui11) : (!firrtl.uint<1>, !firrtl.uint<17>, !firrtl.uint<11>) -> !firrtl.uint<17>
+
+  // Unknown result width should prevent padding.
+  // CHECK: %4 = firrtl.mux(%cond, %ui11, %ui) : (!firrtl.uint<1>, !firrtl.uint<11>, !firrtl.uint) -> !firrtl.uint
+  // CHECK: %5 = firrtl.mux(%cond, %ui, %ui11) : (!firrtl.uint<1>, !firrtl.uint, !firrtl.uint<11>) -> !firrtl.uint
+  %2 = firrtl.mux(%cond, %ui11, %ui) : (!firrtl.uint<1>, !firrtl.uint<11>, !firrtl.uint) -> !firrtl.uint
+  %3 = firrtl.mux(%cond, %ui, %ui11) : (!firrtl.uint<1>, !firrtl.uint, !firrtl.uint<11>) -> !firrtl.uint
+
+  // Padding to equal width operands should enable constant-select folds.
+  // CHECK: %6 = firrtl.pad %ui11, 17 : (!firrtl.uint<11>) -> !firrtl.uint<17>
+  // CHECK: %7 = firrtl.pad %ui11, 17 : (!firrtl.uint<11>) -> !firrtl.uint<17>
+  // CHECK: firrtl.connect %z, %ui17 : !firrtl.uint, !firrtl.uint<17>
+  // CHECK: firrtl.connect %z, %6 : !firrtl.uint, !firrtl.uint<17>
+  // CHECK: firrtl.connect %z, %7 : !firrtl.uint, !firrtl.uint<17>
+  // CHECK: firrtl.connect %z, %ui17 : !firrtl.uint, !firrtl.uint<17>
+  %4 = firrtl.mux(%c0_ui1, %ui11, %ui17) : (!firrtl.uint<1>, !firrtl.uint<11>, !firrtl.uint<17>) -> !firrtl.uint<17>
+  %5 = firrtl.mux(%c0_ui1, %ui17, %ui11) : (!firrtl.uint<1>, !firrtl.uint<17>, !firrtl.uint<11>) -> !firrtl.uint<17>
+  %6 = firrtl.mux(%c1_ui1, %ui11, %ui17) : (!firrtl.uint<1>, !firrtl.uint<11>, !firrtl.uint<17>) -> !firrtl.uint<17>
+  %7 = firrtl.mux(%c1_ui1, %ui17, %ui11) : (!firrtl.uint<1>, !firrtl.uint<17>, !firrtl.uint<11>) -> !firrtl.uint<17>
+
+  firrtl.connect %z, %0 : !firrtl.uint, !firrtl.uint<17>
+  firrtl.connect %z, %1 : !firrtl.uint, !firrtl.uint<17>
+  firrtl.connect %z, %2 : !firrtl.uint, !firrtl.uint
+  firrtl.connect %z, %3 : !firrtl.uint, !firrtl.uint
+  firrtl.connect %z, %4 : !firrtl.uint, !firrtl.uint<17>
+  firrtl.connect %z, %5 : !firrtl.uint, !firrtl.uint<17>
+  firrtl.connect %z, %6 : !firrtl.uint, !firrtl.uint<17>
+  firrtl.connect %z, %7 : !firrtl.uint, !firrtl.uint<17>
```

Comment:
This is fine to leave in, but I will normally remove it to keep the test case smaller. (Or maybe this is needed to prevent something from DCE'd away?)

Reply:
Yeah the canonicalizer DCEs these away. I guess it does a quick use-list-empty check after canonicalization, and drops things if they're not used.